Manager/Manager.php

```diff
@@ -78,6 +89,32 @@ public function save($entity, array $options = array())
      */
     public function remove($entity)
     {
-        $this->remover->remove($entity);
+        if (!$this->isLinkedToProduct($entity)) {
+            $this->remover->remove($entity);
+        }
+    }
+
+    /**
+     * Check if the entity is linked to one or more products
+     *
+     * @param ReferenceDataInterface|object $entity
+     *
+     * @return bool
+     */
+    protected function isLinkedToProduct($entity)
```

Comment:
I would extract all this method in a dedicated checker/validator.
Easier to test, less BC :)

Reply:
:+1: